### ISSUE TITLE
Multiple Context-Roots with Jaxrs

### ISSUE THREAD

**Issue by Fival85** [stoicflame/enunciate#275]
Hello,

we use at least one year enunciate and it is a great framework to document software :) 

But we still have problems with the links in the generated documention. They are still wrong, because we use a jboss-service with many services. So each service has its own context-root. 
Enunciate (1 and 2) use a single context root for every documentated service, but we need different roots for different services.

So i implement an annotation "ServiceContextRoot" in the enunciate-framework and changes the module "jaxrs", to process this annotation. A REST-Facade/Api with this annotation could set his own root-context.

I dont test every scenario, so i couldnt promise that the patch is release-ready, but for our purposes it is sufficient.

I have no rights to push to your repository, so i couldn't make a pull request.  I changed only the class "JaxrsModule" and added the "ServiceContextRoot"-Annotation in the enunciate-core-annotations module (and the version-numbers, but you could ignore this. The code-base is the tag 2.2.0)

Anyway i think its a good idea to have such an annotation to set the context root hard. But i think too, that my implementation isn't the best way. I hope you could reuse my idea or code. So if you like, we could send you the changed code or make a request, if the rights are correct :)

Greetings
Dennis Goermann

**Comment by stoicflame** [comment]
Hi.

That seems like a reasonable enhancement. I'd be very interested in seeing a pull request. You can still prepare a pull request even though you don't have push rights to the repo. See:

https://help.github.com/articles/using-pull-requests/

Let's start with that and go from there.

**Comment by Fival85** [comment]
Hi again, 

thank you for your response. I have created a pull request, but i think you could only reuse the idea. The code is perfect for outher purposes, but not for generally... :)

Greetings
Dennis

**Comment by stoicflame** [comment]
Okay, thank you again for the pull request. I can now see what you've done for your purposes.

The concept is solid, but it needs to be refactored a bit before it gets pushed to the main branch. I've committed the relevant changes to a new branch called "multiple-context-roots" where we can work on it from there.

There are multiple problems with the proposed patch, including:

* Client-side code documentation engine can't find a good example resource method to invoke.
* WADL generation is broken; it's not taking into account the multiple context roots.
* `@ApplicationPath` annotation is getting used differently, likely breaking a lot of consumers who use it.

So before I know how to best move forward, I need to ask about the requirements. Is the only purpose of this enhancement request to fix the broken links to the resources from the generated docs and the Swagger UI? Or do you also want the generated documentation to be grouped differently, per service context? I'm wondering if we can get away with simply enhancing the functionality of the `@ResourceGroup` annotation to apply a different path? I think that way would be easier to implement, but it would require you to use "annotation-based" resource grouping in order to get the documentation grouping benefits. (See the [`groupBy` attribute](https://github.com/stoicflame/enunciate/wiki/Module-JAXRS#configuration) of the JAX-RS Module.)

So instead of using a new `@ServiceContext` annotation, you'd use the `@ResourceGroup` annotation, like this:

```
@ResourceGroup(
  context = "/subcontext"
)
@Path("/my/resource")
public class MyResource {}
```

And that annotation would need to be applied to all your resource classes.

Would that meet your needs?

**Comment by Fival85** [comment]
Hello again and thank you for your response.

We only want to repair the broken links. 

Wether the @ResourceGroup-Annotation is the best way, i couldn't say. I think from the perspective of REST it is a good way, but from the perspective of a service, it make more sense if the context-root is a single annotation.

Anyway: I think enunciate should support an annotation to set the base-url/context-root hard to a specifiec value per REST-API.

By the way: If you want to support a JBOSS-Server. The context-root of a service will be determine by the pom.xml of the ear-Project:
`	<plugin>
				<artifactId>maven-ear-plugin</artifactId>
				<version>${maven-ear-plugin.version}</version>
				<configuration>
					<defaultLibBundleDir>lib</defaultLibBundleDir>
					<displayName>usage-service2</displayName>
					<applicationName>usage-service2</applicationName>
					<version>6</version>
					<modules>
						<ejbModule>
							<groupId>de.mobilcom.usage</groupId>
							<artifactId>usage-service-ejb</artifactId>
							<bundleFileName>usage-service-ejb.jar</bundleFileName>
						</ejbModule>
                        <webModule>
                            <groupId>${project.groupId}</groupId>
                            <artifactId>${project.parent.artifactId}-war</artifactId>
                            <bundleFileName>${project.parent.artifactId}-war.war</bundleFileName>
                            <contextRoot>${project.parent.artifactId}</contextRoot>
                        </webModule>
					</modules>
					<archive>
						<manifest>
							<addClasspath>true</addClasspath>
						</manifest>
						<index>false</index>
						<manifestEntries>
							<Version>${project.version}</Version>
						</manifestEntries>
					</archive>
					<generateApplicationXml>true</generateApplicationXml>
				</configuration>
			</plugin>`

Greetz Dennis

**Comment by stoicflame** [comment]
Okay.

At b1646752139a52061639dc91b019b6d1f39e4767 I modified your original design. I kept the `@com.webcohesion.enunciate.metadata.rs.ServiceContextRoot` annotation, but changed the element name from `context()` to `value()`. I also rolled back the changes you made to the JaxrsModule and pushed the calculation of the context root down to the resource grouping logic.

The new annotation [has been documented on the wiki](https://github.com/stoicflame/enunciate/wiki/Enunciate-Specific-Annotations#cwemrsservicecontextroot).
